### PR TITLE
scala notes

### PR DESCRIPTION
**Pull request by olivierschultz**
* error handling
* category theory
* functor
* applicative
* monad
* hkt
* algebra: ADT
* typeclass

### CODE REVIEW

File: error_handling.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+```
+def forecast(city: String): Either[Error, Double] =
+  for {
+  	location <- lookupCity(city)
+  	forecast <- weather(location)
+  	temp <- getTemperature(forecast).toEither("No temp")
+  } yield (forecast)
+```
+
+Monadic types(`Either`, `Option`) that are representing error situations shortcut on error. Once there there is a `None` or `Left`, the other rows in the sequence will not get called and the error value gets passed to yield.
```
Comment (BerthelotAnthony) [inline error_handling.md]: Remove one **there**

Comment (BerthelotAnthony) [inline error_handling.md]: Maybe: `and the value **will get** passed`

File: error_handling.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+def forecast[F[_]: Monad : MonadError](city: String): F[Either[Error, Double] =
+  for {
+  	location <- EitherT(lookupCity(city))
+  	forecast <- EitherT(weather(location))
+  	temp <- EitherT.fromEithergetTemperature(forecast).toEither("No temp")
+  } yield (forecast)
+```
+
+_Note_:
+  - You can also improve th readibility by adding implicit classes
```
Comment (BerthelotAnthony) [inline error_handling.md]: **the** | readability

File: scala/category_theory.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
+Category Theory in Scala
+========================
+
+## What's a category?
+
+### Category
+
+A `category` is made up from `objects` and `maps` (aka `morphisms` or `arrows`) between these objects. Maps can be `composed` in an `associative` fashion and for each object there is an `identity map` which is neutral with regard to composition.
```
Comment (BerthelotAnthony) [inline scala/category_theory.md]: Think it's with regard**s** to

File: scala/category_theory.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+### Associativity
+
+Let's add another map h: C -> D. Then the associativity law requires the following to hold true:
+
+`(h ο g) ο f = h ο (f ο g)`
+
+### Identity
+
+THese require that for each object X of the category there is an identity map `1X: X → X` which satisfies the following rules:
```
Comment (BerthelotAnthony) [inline scala/category_theory.md]: These

Comment (BerthelotAnthony) [reply scala/category_theory.md]: requires

File: scala/category_theory.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  	"satisfy identity" in {
+  	  Prop forAll { (i: Int) =>
+  	    compose(h, id[Int])(i) mustEqual compose(id[Int], h)(i) 
+  	  } must pass	
+  	}
+  }	
+}
+```
+
+Let's move along to a more general category in Scala.
```
Comment (BerthelotAnthony) [inline scala/category_theory.md]: missing words ? Let's move along to ... a more

Comment (olivierschultz) [reply scala/category_theory.md]: rephrased with Let's write a more generic category in Scala.

File: scala/evaluation.md [hunk 6]
```diff
@@ -17,7 +17,10 @@ _Note_: One good reason for choosing `def` over `val`, especially in abstract cl
 
 ## lazy val
 
-`lazy val` is evaluated once it gets accessed.
+`lazy val` is evaluated once it gets first accessed.
 
-__source__: https://stackoverflow.com/questions/9449474/def-vs-val-vs-lazy-val-evaluation-in-scala
+_note: `lazy val` other than a regular `val` has to pay the cost of checking the initialization state of each eaccess._
```
Comment (BerthelotAnthony) [inline scala/evaluation.md]: access ?

File: scala/algebra.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+In that case:
+  - symbols: `Amount`, `Price`, `ID`, `Item`, `Sales`
+  - operations: `combineAll`, `generateReport`
+  - properties and rules: the tests
+
+## Algebraic Data Type
+
+ADT are just a way of modelling data, a type formed by combining other types, in terms of two patterns:
+  - logical `OR`, also called sum type, such as `Either`, which is `Left` or `Right`. It's done in Scala by using `sealed trait` and their instances as case classes or case object.
```
Comment (BerthelotAnthony) [inline scala/algebra.md]: or case object**S**

File: scala/algebra.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  - logical `AND`, also called product type, such as `::`, which is a `head` and a `tail`. It's done ine Scala by using `case class`
```
Comment (BerthelotAnthony) [inline scala/algebra.md]: in

File: scala/error_handling.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+```scala
+def forecast(city: String): Either[Error, Double] =
+  for {
+  	location <- lookupCity(city)
+  	forecast <- weather(location)
+  	temp <- getTemperature(forecast).toEither("No temp")
+  } yield (forecast)
+```
+
+Monadic types that are representing error situations shortcut on error. Once there is a `None` or `Left`, the other rows in the sequence will not get called and the  value will get passed to yield.
```
Comment (BerthelotAnthony) [inline scala/error_handling.md]: remove **that**

File: scala/error_handling.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+`EitherT[F, A, B]` wraps a value f of type F[Either[A, B]].
+
+```
+def forecast[F[_]: Monad : MonadError](city: String): F[Either[Error, Double] =
+  for {
+  	location <- EitherT(lookupCity(city))
+  	forecast <- EitherT(weather(location))
+  	temp <- EitherT.fromEithergetTemperature(forecast).toEither("No temp")
+  } yield (forecast)
```
Comment (BerthelotAnthony) [inline scala/error_handling.md]: add scala syntax color ?

File: scala/functional_programming.md [hunk 11]
```diff
@@ -65,7 +42,25 @@ def add(a: Int, b: Int): Int = a + b
 
 _Note_: A function that returns a __Unit__ is usually a hint that it's performing a side effect.
 
-Functions that respect those rules are called: _Pure Function_.
+## Referential Transparency
+
+If a function satisfy those properties you ge referential transparency. Pure functions calls can be replacable by their results anywhere they are used without altering the rest of the program.
```
Comment (BerthelotAnthony) [inline scala/functional_programming.md]: satisfies

Comment (BerthelotAnthony) [inline scala/functional_programming.md]: you ge? you **get**

File: scala/functional_programming.md [hunk 12]
```diff
@@ -65,7 +42,25 @@ def add(a: Int, b: Int): Int = a + b
...
+  s.toUpperCase 
+}
+```
+
+```
+// Referential transparent
+def helloWorld(s: String): String = {
+  s.toUpperCase 
+}
+```
```
Comment (BerthelotAnthony) [inline scala/functional_programming.md]: scala color syntax ?

**Pull request merged**